### ISSUE TITLE
skipAllAfterTimeLimit throws an error instead of skipping tests

### ISSUE THREAD

**Issue by ZeRego** [dubzzz/fast-check#423]
## 🐛 Bug Report

skipAllAfterTimeLimit throws an error instead of skipping the remaining tests like the name suggests

## To Reproduce

Steps to reproduce:

```describe('skipAllAfterTimeLimit', () => {
    it(‘should skip tests’, async () => {
        jest.useRealTimers();
        const property = fc.asyncProperty(
            fc.constant(''),
            async () => {
                await new Promise(resolve => setTimeout(resolve, 100));
                expect(true).toBe(true);
            }
        );
        await fc.assert(property, {
            numRuns: 10,
            skipAllAfterTimeLimit: 1,

        });
    });
});
```

## Expected behavior

It should skip all remaining tests instead of throwing error. 
We could have a parameter, similar to maxSkipsPerRun, to specify the max amount of tests it can skip.

## Your environment

| Packages / Softwares  | Version(s) |
| --------------------- | ---------- |
| fast-check            |       1.16.2     |
| node                  |     10.14.1       |
| TypeScript          |     3.4.5       |

**Comment by dubzzz** [comment]
@ZeRego Thanks a lot for the report. I indeed reproduced the issue immediately. I fixed _part_ of the issue and added the tests to avoid this issue to come back in the future.

---

I initially added the flag `skipAllAfterTimeLimit` for very long model based tests applied onto UI. The idea was just to avoid running into _infinite_ shrinks in my CI (maybe hours of shrink when I launched it on really complex sets of commands). After this delay, instead of running the tests, fast-check does as if you shrunk but marked the test with precondition failure (as `fc.pre` does).

Basically the feature adds a precondition on top of all the properties:
```js
fc.pre(notExpired());
// here the property code
```

---

The need you have seem to be a bit different: you want to stop the run where it is and report its curent status. I definitely have to add such feature.

**Comment by dubzzz** [comment]
I'll release the fix into 1.15.x and 1.16.x as soon as possible, so that the feature does what it says. As told above I'll need to add another feature that just stops where the runner is and reports its current status (no matter if it still have to generate test cases to fulfill `numRuns` constraint).

**Comment by dubzzz** [comment]
I double-checked the code this morning, the feature you asked should be quite easy to add.

_As `skipAllAfterTime` was designed to skip runs (using `fc.pre`), I prefer to have it as another flag. Maybe `stopAfterTime` or `interruptAfterTime`. Anyway it will be very important to highlight that this feature can stop the execution before we reach the number of test cases_

I can either build it on my side or send you pointers on where to go.

**Comment by dubzzz** [comment]
@ZeRego I drafted a PR to implement a `interruptAfterTimeLimit` feature (code is a pure draft, it would need some more tests). I still have opened questions around it.

**Comment by dubzzz** [comment]
@ZeRego What's your feeling concerning `interruptAfterTimeLimit`?

**Comment by ZeRego** [comment]
Sorry for the late response ( didn't get any notification from github ).
I would be more inclined to switch the names since I associate `skip` to a graceful fallback and `interrupt` to a unexpected failure.

> As skipAllAfterTime was designed to skip runs (using fc.pre)

In the current code `skipAllAfterTime` is skipping runs by throwing an error. Which feels more like an interrupt.
On the opposite side, `interruptAfterTimeLimit`, actually skips the remaining numRuns .

**Comment by dubzzz** [comment]
**skipAllAfterTimeLimit**

I agree that `skipAllAfterTimeLimit` was not well defined. The initial idea behind this feature was: mark runs as skipped as soon as the time limit expired.

In fast-check, a run can have 3 different states: success, failure, skipped (mainly triggered by `fc.pre`).
Initially `skipAllAfterTimeLimit` was supposed to mark all runs as skipped after the time limit.

There are two possible outcomes:
- timeout occurs during shrinking: shrinking will be interrupted by marking all the shrunk values as skipped
- timeout occurs while there are remaining values to generate: test will fail because it will never be able to reach the required number of tests

See https://github.com/dubzzz/fast-check/pull/425 which just fix the feature as it was initially defined.

**interruptAfterTimeLimit**

On the contrary, `interruptAfterTimeLimit` does not mark runs as skipped. It just stops the execution of the test suite. There two possibilities: either it should be marked as a failure, or a success with not enough runs (see `markInterruptAsFailure` in https://github.com/dubzzz/fast-check/pull/428).

Internally `interruptAfterTimeLimit` really interrupts the execution either by marking it as failure or as success. While `skipAllAfterTimeLimit` just marks everything as skipped.

**conclusion**

- `skipAllAfterTimeLimit`: mark runs as skipped after timeout (shrunk values will still be produced)
- `interruptAfterTimeLimit`: stop execution after timeout - test is considered as success if it has never encountered an issue
- `interruptAfterTimeLimit` and `markInterruptAsFailure:true`: stop execution after timeout - an interrupted test is considered as a failed test


Actually I'm still looking for the right way to add this interrupt/skip/whatever feature (maybe replacing `skipAllAfterTimeLimit`).

**Comment by ZeRego** [comment]
Thanks for the explanation.
I agree with your approach for `interruptAfterTimeLimit` and `markInterruptAsFailure`. ( will try to find some time and help you with the feature branch )
I think we can close this issue since is was for the bug in `skipAllAfterTimeLimit`.

**Comment by ZeRego** [comment]
I see that the fail message with very verbose mode only shows the generated value.
Could we add the seed and path to each one ? 

Current error message:
```
Error: Failed to run property, too many pre-condition failures encountered

Ran 9 time(s)
Skipped 1001 time(s)

Execution summary:
√ [""]
√ [""]
√ [""]
√ [""]
√ [""]
√ [""]
√ [""]
√ [""]
√ [""]
! [""]
! [""]
...
! [""]

Hint (1): Try to reduce the number of rejected values by combining map, flatMap and built-in arbitraries
Hint (2): Increase failure tolerance by setting maxSkipsPerRun to an higher value
```

**Comment by dubzzz** [comment]
Good catch for the missing seed 👍
It should be easy to fix. If you're interested, I can provide you with some pointers to do so.

**Comment by dubzzz** [comment]
What's the best option? `markInterruptAsFailure` or `markInterruptAsSuccess`? 🤔

**Comment by dubzzz** [comment]
> I think we can close this issue since is was for the bug in skipAllAfterTimeLimit.
Closing the issue, please feel free to re-open or open a new issue if you detect other issues
